### DISCUSSION TITLE
What's best practice of importing sub-modules from an aliased module?

### DISCUSSION THREAD

**Discussion by Azureblade3808** [microsoft/pyright#3631] [General]
I was using RxPY3 (package name: `rx`) and is trying to migrate to RxPY4 (package name: `reactivex`).

With RxPY3, it's common that a sub-module of `rx` is imported, such like
```python
import rx, rx.operators, rx.subject

a = rx.subject.BehaviorSubject(0)
b = a.pipe(rx.operators.distinct_until_changed())
```
. And now with RxPY4, I've tried
```python
import reactivex as rx, reactivex.operators, reactivex.subject

a = rx.subject.BehaviorSubject(0)
b = a.pipe(rx.operators.distinct_until_changed())
```
to keep the code modification minimized, but see that it is not supported by Pyright.

I'm currently writing
```python
import reactivex, reactivex.operators, reactivex.subject

rx = reactivex

a = rx.subject.BehaviorSubject(0)
b = a.pipe(rx.operators.distinct_until_changed())
```
to get the work done. What bugs me a little is that the alias line may get separated far from the import line.

It's not a general case and I believe that `reactivex.operators` should be automatically imported with `reactivex` so that this situation could totally be avoided. But I'd like to see what you would do with this case.

**Comment by erictraut** [comment]
Your approach in the third code snippet is what I would recommend.

Pyright intentionally doesn't model implicit side effects of the import resolver because they are generally fragile assumptions. Each import is intentionally modeled as though it is performed in isolation.
